File: R/occlusion.R
#' What part of the record a fitted model reads
#'
#' Holds one bin of the record back at a time, rescores the held-out units, and records the fall in
#' score as that bin's weight. Nothing is refitted: the models kept by
#' `grain_ladder(keep_fits = TRUE)` are the ones read, so the profile describes the models that
#' produced the reported scores rather than a fresh set of them.
#'
#' A model has to be shown something in place of a held-back bin, and what it is shown decides what
#' the weight means. Permuting the bin's values across units keeps the observed readings exactly
#' and cuts only the link between a reading and its unit. Replacing every unit by the fitting-fold
#' mean removes all between-unit variation while keeping the shape of the year. Replacing the bin
#' by each unit's own mean over the record keeps how warm a unit is and removes only that bin's
#' departure from it.
#'
#' Read with `over = "channel"` the same machinery asks what each statistic of a grain carries,
#' holding one channel back across the whole record instead of one bin across all channels.
#'
#' @param x A [timesift()] fit, or a [grain_ladder()] result, in either case fitted with
#'   `keep_fits = TRUE`.
#' @param data The representation set the ladder was fitted on.
#' @param y The response it was fitted to.
#' @param arm The arm to read, as `"grain|learner"` or `"learner"`.
#' @param ... Passed to the method.
#' @param over `"bin"` to hold each bin back in turn, `"channel"` for each channel.
#' @param substitute What a held-back part is replaced by: `"permute"`, `"fold_mean"` or
#'   `"unit_mean"`.
#' @param metric Name of the registered metric the rescoring is read by, or a function of
#'   `(y, p)`. Left unset it is the one the fit was scored under, so a weight is a fall in the
#'   number [summary()] reports rather than in a second one. `"roc_auc"` is usually the steadier
#'   reading over many rescorings: it responds to every reordering of the units, where a maximum
#'   over thresholds frequently does not move at all.
#' @param permutations Draws averaged over, for `substitute = "permute"`.
#' @param seed Random seed.
#'
#' @return A data frame of one row per held-back part and variable, carrying the mean weight over
#'   folds and the score with and without the part.
#'
#' @examplesIf requireNamespace("glmnet", quietly = TRUE)
#' set.seed(1)
#' t <- seq(as.POSIXct("2021-09-01", tz = "UTC"), by = "hour", length.out = 24 * 120)
#' units <- sprintf("p%02d", 1:40)
#' warmth <- rnorm(40)
#' d <- data.frame(
#'   plot = rep(units, each = length(t)), t = rep(t, length(units)),
#'   temp = as.numeric(vapply(warmth, function(w) w + sin(seq_along(t) / 300) + rnorm(length(t)),
#'                            numeric(length(t)))))
#' y <- matrix(rbinom(80, 1, plogis(c(warmth, -warmth))), nrow = 40,
#'             dimnames = list(units, c("sp1", "sp2")))
#' x <- grain_matrix(d, plot, t, temp, grain = "month")
#' lad <- grain_ladder(x, y, elasticnet(), folds = fold_map(y, v = 3),
#'                      keep_fits = TRUE, verbose = FALSE)
#' head(occlusion(lad, x, y, "month|elasticnet", permutations = 3))
#'
#' @export
occlusion <- function(x, ...) UseMethod("occlusion")

#' @rdname occlusion
#' @export
occlusion.default <- function(x, ...) {
  stop("expected a timesift() run or a grain_ladder() result, got ", class(x)[1L], ".",
       call. = FALSE)
}

#' @rdname occlusion
#' @export
occlusion.timesift_ladder <- function(x, data, y, arm, over = c("bin", "channel"),
                                     substitute = c("permute", "fold_mean", "unit_mean"),
                                     metric = NULL, permutations = 20L, seed = 1L,
                                     ...) {
  ladder <- x
  x <- data
  over <- match.arg(over)
  substitute <- match.arg(substitute)
  fits <- attr(ladder, "fits")
  if (is.null(fits)) {
    stop("this ladder kept no fits. Refit with grain_ladder(..., keep_fits = TRUE).",
         call. = FALSE)
  }
  rows <- .arm_rows(ladder, arm)
  label <- attr(rows, "label")
  grain <- attr(rows, "grain")
  set <- .as_set(x)
  m <- set[[grain]]
  if (is.null(m)) {
    stop("the representation carries no \"", grain, "\" grain.", call. = FALSE)
  }
  units <- dimnames(m)[[1L]]
  y <- .align_response(.responses_reg$get(attr(ladder, "response"))$prepare(y), units)
  f <- .as_folds(attr(ladder, "folds"), units)
  cells <- attr(ladder, "cells")
  # Left unset the profile is read by the metric the fit was scored under, so a weight is a fall
  # in the number the summary reports rather than in a second one.
  metric <- if (is.null(metric)) list(fn = attr(ladder, "scorer"), name = attr(ladder, "metric"))
            else .as_metric(metric)
  score <- metric$fn

  parts <- if (over == "bin") seq_len(dim(m)[2L]) else seq_len(dim(m)[3L])
  labels <- if (over == "bin") dimnames(m)[[2L]] else dimnames(m)[[3L]]

  old <- .seed_state()
  on.exit(.restore_seed(old), add = TRUE)
  set.seed(seed)

  out <- list()
  for (k in sort(unique(f))) {
    fit <- fits[[paste(label, k, sep = "|")]]
    if (is.null(fit)) {
      next
    }
    test <- which(f == k)
    train <- which(f != k)
    base <- stats::predict(fit, .subset_units(m, test))
    ok <- .scorable_for(cells, colnames(y), k)
    full <- .score_columns(y[test, , drop = FALSE], base, ok, score)

    for (i in seq_along(parts)) {
      draws <- if (substitute == "permute") permutations else 1L
      acc <- matrix(0, nrow = draws, ncol = ncol(y))
      for (r in seq_len(draws)) {
        occluded <- .occlude(m, test, train, parts[i], over, substitute)
        acc[r, ] <- .score_columns(y[test, , drop = FALSE],
                                   stats::predict(fit, occluded), ok, score)
      }
      out[[length(out) + 1L]] <- data.frame(
        part = labels[i], variable = colnames(y), fold = k,
        score_full = full, score_held_back = colMeans(acc),
        weight = full - colMeans(acc), stringsAsFactors = FALSE)
    }
  }
  out <- do.call(rbind, out)
  out <- out[!is.na(out$weight), , drop = FALSE]
  agg <- stats::aggregate(out[c("score_full", "score_held_back", "weight")],
                          out[c("part", "variable")], mean)
  agg$part <- factor(agg$part, levels = labels)
  agg <- agg[order(agg$part, agg$variable, method = "radix"), ]
  agg$part <- as.character(agg$part)
  rownames(agg) <- NULL
  structure(agg, class = c("timesift_occlusion", "data.frame"), arm = label, over = over,
            substitute = substitute, metric = metric$name)
}

.occlude <- function(m, test, train, i, over, substitute) {
  sub <- .subset_units(m, test)
  n <- dim(sub)[1L]
  b <- dim(sub)[2L]
  if (over == "channel") {
    sub[, , i] <- switch(
      substitute,
      permute = .plane(sub, i)[sample.int(n), , drop = FALSE],
      fold_mean = matrix(colMeans(.plane(m, i)[train, , drop = FALSE]),
                         nrow = n, ncol = b, byrow = TRUE),
      unit_mean = matrix(rowMeans(.plane(sub, i)), nrow = n, ncol = b)
    )
    return(sub)
  }
  for (ch in seq_len(dim(sub)[3L])) {
    sub[, i, ch] <- switch(
      substitute,
      permute = sub[sample.int(n), i, ch],
      fold_mean = mean(m[train, i, ch]),
      unit_mean = rowMeans(.plane(sub, ch))
    )
  }
  sub
}

# One channel of a representation as a [unit, bin] matrix, whatever the number of bins. Dropping to
# a vector at a single bin is what makes the row and column means below silently wrong.
.plane <- function(x, channel) {
  matrix(as.numeric(x[, , channel]), nrow = dim(x)[1L], ncol = dim(x)[2L])
}

.scorable_for <- function(cells, variables, fold) {
  cells$scorable[match(paste(variables, fold), paste(cells$variable, cells$fold))]
}

.score_columns <- function(y, p, ok, score) {
  vapply(seq_len(ncol(y)), function(j) {
    if (isTRUE(ok[j])) score(y[, j], p[, j]) else NA_real_
  }, numeric(1L))
}

#' @export
print.timesift_occlusion <- function(x, ...) {
  cat("<timesift occlusion>", attr(x, "arm"), "read by", attr(x, "metric"), "\n")
  cat("held back:", attr(x, "over"), "; substitute:", attr(x, "substitute"), "\n")
  w <- stats::aggregate(list(weight = x$weight), x["part"], mean)
  w <- w[order(-w$weight), ]
  cat("heaviest:", paste(utils::head(w$part, 5L), collapse = ", "), "\n")
  print(utils::head(as.data.frame(x), 6L))
  invisible(x)
}

File: python/timesift/occlusion.py
"""What part of the record a fitted model reads, and what an already-reduced feature table is."""

from __future__ import annotations

from dataclasses import replace

import numpy as np

from .ladder import _best_arm, _split_arm
from .registry import RESPONSES, resolve_metric
from .representation import TimesiftMatrix
from .response import Folds, Response, as_response

__all__ = ["feature_matrix"]


def feature_matrix(m, units=None, features=None, label: str = "features") -> TimesiftMatrix:
    """Bring an already-reduced feature table into a ladder as a one-channel representation.

    It carries no time axis, because it has none: the reduction already happened, elsewhere, and
    what reaches the model is a list of numbers per unit. That is the whole point of comparing
    against it.
    """
    m = np.asarray(m, dtype=np.float64)
    n_u, n_f = m.shape
    units = tuple(units) if units is not None else tuple(str(i + 1) for i in range(n_u))
    features = tuple(features) if features is not None else tuple(f"f{j}" for j in range(n_f))
    empty = np.full(n_f, np.datetime64("NaT"), dtype="datetime64[s]")
    return TimesiftMatrix(values=m.reshape(n_u, n_f, 1), units=units, bins=features,
                        stats=(label,), grain=label, year_start="", bin_start=empty,
                        bin_end=empty, bin_n=np.zeros((n_u, n_f), dtype=np.int64),
                        bin_partial=np.zeros(n_f, dtype=bool))


def ladder_occlusion(ladder, x, y: Response, arm: str, over: str = "bin",
                  substitute: str = "permute", metric=None,
                  permutations: int = 20, seed: int = 1):
    """Hold one bin of the record back at a time and record the fall in score as its weight.

    Nothing is refitted: the models kept by ``grain_ladder(keep_fits=True)`` are the ones read.
    What a held-back bin is replaced by decides what the weight means, so the substitute is part of
    the answer: ``permute`` keeps the observed readings and cuts only the link between a reading
    and its unit, ``fold_mean`` removes all between-unit variation, ``unit_mean`` keeps how warm a
    unit is and removes only that bin's departure from it. A bin is held back whole, every channel
    of it moved by one permutation, and a channel that is the same for every unit, as the calendar
    channels are, is left in place.
    """
    if not ladder.fits:
        raise ValueError("this ladder kept no fits; refit with grain_ladder(..., keep_fits=True)")
    grain, learner = _split_arm(ladder, _best_arm(ladder, arm))
    if isinstance(x, TimesiftMatrix):
        m = x
    else:
        m = dict(x).get(grain)
        if m is None:
            raise ValueError(f'the representation carries no "{grain}" grain')
    prefix = f"{grain}|{learner}|"
    fits = {int(key[len(prefix):]): fit for key, fit in ladder.fits.items()
            if key.startswith(prefix)}
    if not fits:
        raise KeyError(f'this ladder kept no fits for the arm "{grain}|{learner}"')
    # Left unset the profile is read by the metric the fit was scored under, so a weight is a fall
    # in the number the summary reports rather than in a second one.
    return occlusion_profile(fits, m, y, ladder.folds, over=over, substitute=substitute,
                             metric=ladder.scorer if metric is None else metric,
                             response=ladder.response, permutations=permutations, seed=seed)


def occlusion_profile(fits: dict, m: TimesiftMatrix, y, folds, over: str = "bin",
                      substitute: str = "permute", metric="roc_auc",
                      response: str = "presence_absence",
                      permutations: int = 20, seed: int = 1):
    """The occlusion itself: one fitted model per fold, read on the units that model held out.

    A ladder reaches it through ``ladder_occlusion`` and a fitted ``timesift`` through ``occlusion``,
    so what a withheld bin costs has one definition whichever door it is asked through.
    """
    if over not in ("bin", "channel"):
        raise ValueError(f"`over` must be 'bin' or 'channel', got {over!r}")
    if substitute not in ("permute", "fold_mean", "unit_mean"):
        raise ValueError(f"unknown substitute {substitute!r}")

    # The response reaches its own head's prepare, as it does everywhere else, so a head that is
    # not presence-absence can be occluded too.
    spec = RESPONSES.get(response)
    y = spec["prepare"](as_response(y)).align(m.units)
    folds = Folds.coerce(folds, m.units).align(m.units)
    f = folds.fold
    score = resolve_metric(metric)[0]
    # The cells a score is defined on, from the response and the fold map alone, so a weight is read
    # on the cells the ladder scored and nowhere else.
    cells = spec["cells"](y, folds)
    scorable = {(v, int(k)): bool(ok) for v, k, ok in zip(cells.variable, cells.fold,
                                                           cells.scorable)}
    held = _unit_varying(m.values)

    n_parts = m.values.shape[1] if over == "bin" else m.values.shape[2]
    labels = m.bins if over == "bin" else m.stats
    rng = np.random.default_rng(seed)

    weight = np.full((n_parts, len(y.variables)), np.nan)
    counted = np.zeros((n_parts, len(y.variables)))
    for k in np.unique(f):
        fit = fits.get(int(k))
        if fit is None:
            continue
        test = np.flatnonzero(f == k)
        train = np.flatnonzero(f != k)
        sub = m.take_units(test)
        ok_cell = [scorable.get((v, int(k)), False) for v in y.variables]

        def score_columns(p):
            return np.asarray([score(y.values[test, j], p[:, j]) if ok_cell[j] else np.nan
                               for j in range(len(y.variables))])

        # The baseline is one prediction for the fold, not one per variable: the model reads the
        # whole block and an encoder would otherwise be rebuilt from its arrays once per response.
        base = score_columns(fit.predict(sub))
        for i in range(n_parts):
            draws = permutations if substitute == "permute" else 1
            acc = np.zeros((draws, len(y.variables)))
            for r in range(draws):
                occluded = _occlude(m, sub, train, i, over, substitute, rng, held)
                acc[r] = score_columns(fit.predict(occluded))
            fall = base - acc.mean(axis=0)
            ok = np.isfinite(fall)
            weight[i, ok] = np.where(np.isnan(weight[i, ok]), 0, weight[i, ok]) + fall[ok]
            counted[i, ok] += 1

    with np.errstate(invalid="ignore"):
        weight = weight / np.where(counted > 0, counted, np.nan)
    return {"part": list(labels), "variable": list(y.variables), "weight": weight}


def _unit_varying(values: np.ndarray) -> np.ndarray:
    """The channels that differ between units somewhere in the record. The rest are the same for
    every unit, as the calendar channels are, and holding a bin back leaves them where they are."""
    return np.flatnonzero((values != values[:1]).any(axis=(0, 1)))


def _occlude(m: TimesiftMatrix, sub: TimesiftMatrix, train, i, over, substitute, rng,
             held) -> TimesiftMatrix:
    values = sub.values.copy()
    n = values.shape[0]
    if over == "channel":
        if substitute == "permute":
            values[:, :, i] = values[rng.permutation(n), :, i]
        elif substitute == "fold_mean":
            values[:, :, i] = m.values[train, :, i].mean(axis=0)[None, :]
        else:
            values[:, :, i] = values[:, :, i].mean(axis=1)[:, None]
        return replace(sub, values=values)
    order = rng.permutation(n) if substitute == "permute" else None
    for ch in held:
        if substitute == "permute":
            values[:, i, ch] = values[order, i, ch]
        elif substitute == "fold_mean":
            values[:, i, ch] = m.values[train, i, ch].mean()
        else:
            values[:, i, ch] = values[:, :, ch].mean(axis=1)
    return replace(sub, values=values)
